### PR TITLE
Swift: Recognize regular expression parse mode flags

### PR DESCRIPTION
**Pull request by geoffw0**
Recognize regular expression parse mode flags in Swift.  For example in `(?i)foo`, the `(?i)` specifies a case-insensitive match of the rest of the regular expression `foo`.

This improves the accuracy of the REDOS and bad HTML filtering queries (and others to come).

Limitations:
1. does not support alternative ways of setting the parse mode, i.e. using function calls / parameters outside of the regex string itself (this will be high priority follow-up work)
2. does not support modes specified in regex literals (because we don't support regex literals yet)
3. does not support setting the mode in the middle of a regex `foo(?i)bar`, which is allowed in Swift
4. does not support setting the mode only for a portion of a regex `(?i:foo)bar`, which is allowed in Swift
5. does not support turning flags off `(?i-s)`, which is allowed in Swift

@erik-krogh @joefarebrother the initial version of this was based on code from Java and is also very similar to Python.  I found that we were failing to recognize multiple mode flags (e.g. `(?is)`), which I fixed with https://github.com/github/codeql/commit/6e80021c4e460be0d7304f18c566eee423b140a4.  Please could you review those changes with a view to both correctness for Swift, and whether or not it would be desirable for me to port that fix to Java and/or Python?

I'm also curious whether we support any of what I described in limitations 3-5 above in other languages.  I suspect we do not, which makes it a pretty low priority for Swift where we're currently playing catch-up - but if we do it would be a starting point to solving them.

### CODE REVIEW

**Comment by erik-krogh** [comment]
The multiple modes commit looks good to me, and I think the other languages could use a port of that fix.

File: swift/ql/lib/codeql/swift/regex/internal/ParseRegex.qll [hunk 1]
```diff
@@ -274,6 +274,58 @@ abstract class RegExp extends Expr {
...
+
+  /**
+   * Gets a mode (if any) of this regular expression. Can be any of:
+   * IGNORECASE
+   * VERBOSE
+   * DOTALL
+   * MULTILINE
+   * UNICODE
+   */
+  string getAMode() { result = this.getModeFromPrefix() }
```
Comment (sashabu) [inline swift/ql/lib/codeql/swift/regex/internal/ParseRegex.qll:327]: What's the advantage of splitting the implementation out into the separate `getModeFromPrefix`?

**Review by sashabu — APPROVED** [review]

File: swift/ql/lib/codeql/swift/regex/internal/ParseRegex.qll (hunk 1 above)
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/regex/internal/ParseRegex.qll:327]: `getModeFromPrefix` gets mode flags set within the regular expression string itself, i.e. the `i` in `(?i)foo`.

`getAMode` gets any mode flags that can affect this regular expression, whether they be set inside the regular expression string or programmatically applied (e.g. due to a call to [Regex.ignoresCase(_:)](https://developer.apple.com/documentation/swift/regex/ignorescase(_:)) at some point before the regex is evaluated).  As yet we only support the former, but you can expect `getAMode` to do more in future PRs.

**Comment by geoffw0** [comment]
Thanks for the review!

**Pull request merged**